macaw-dump/src/Data/Macaw/Dump/Discover.hs

```diff
@@ -0,0 +1,160 @@
+-- addresses from shared libraries are in non-trivial segments). In those cases,
+-- we map to the unstructured 'WPL.Others' with a sufficiently descriptive string.
+--
+-- TODO: import from `Testing`
+posFn :: (MM.MemWidth w) => Text.Text -> MM.MemSegmentOff w -> WPL.Position
+posFn binaryName segOff =
+  case MM.segoffAsAbsoluteAddr segOff of
+    Just mw -> WPL.BinaryPos binaryName (fromIntegral mw)
+    Nothing -> WPL.OtherPos (binaryName <> Text.pack ": " <> Text.pack (show segOff))
+
+-- | Load an ELF file into a macaw 'MM.Memory' (and symbols)
+--
+-- Prints warnings to stderr.
+loadELF ::
+  MML.LoadOptions ->
+  EE.ElfHeaderInfo w ->
+  IO (MM.Memory w, [MEL.MemSymbol w])
+loadELF loadOpts ehi = do
+  case MEL.resolveElfContents loadOpts ehi of
+    Left err -> X.throwIO (ELFResolutionError err)
+    Right (warnings, mem, _mentry, nameAddrList) -> do
+      F.forM_ warnings $ \w -> do
+        IO.hPutStrLn IO.stderr ("WARN: " ++ w)
+      return (mem, nameAddrList)
+
+-- | Run discovery on the provided symbols, or all if none are provided
+runDiscovery ::
+  forall arch w.
+  ( MC.ArchAddrWidth arch ~ w
+  , MC.ArchConstraints arch
+  , MM.MemWidth w
+  ) =>
+  EE.ElfHeaderInfo w ->
+  MAI.ArchitectureInfo arch ->
+  Set.Set BS.ByteString ->
+  IO (MD.DiscoveryState arch)
+runDiscovery headerInfo archInfo symbols = do
+  (mem, nameAddrList) <- loadELF MML.defaultLoadOptions headerInfo
+  let addrSymMap =
+        Map.fromList
+          [ (MEL.memSymbolStart msym, name)
+          | msym <- nameAddrList
+          , let name = MEL.memSymbolName msym
+          , Set.null symbols || Set.member name symbols
+          ]
+  pure (MD.cfgFromAddrs archInfo mem addrSymMap (Map.keys addrSymMap) [])
+
+displayCfgs ::
+  ( MC.ArchConstraints arch
+  , MS.GenArchInfo MS.LLVMMemory arch
+  , CCE.IsSyntaxExtension (MS.MacawExt arch)
+  ) =>
+  FilePath ->
+  MD.DiscoveryState arch ->
+  MS.GenArchVals mem arch ->
+  -- | Also print Crucible CFG?
+  Bool ->
+  IO ()
+displayCfgs path discState archVals printCrucible = do
+  let funInfos = discState Lens.^. MD.funInfo
+  halloc <- CFH.newHandleAllocator
+  F.for_ (Map.toList funInfos) $ \(_addr, Some info) -> do
+    IO.print (PP.pretty info)
+    Monad.when printCrucible $ do
+      let pos = posFn (Text.pack path)
+      let funName =
+            WF.functionNameFromText $
+              Text.decodeUtf8With Text.lenientDecode $
+                MD.discoveredFunName info
+      CCC.SomeCFG ssa <-
+        MS.mkFunCFG (MS.archFunctions archVals) halloc funName pos info
+      IO.print (CCC.ppCFG' True (CAP.postdomInfo ssa) ssa)
+
+data DiscoverConfig = DiscoverConfig
+  { -- Arguments
+    discBinPath :: FilePath
+  , discSymbols :: [BS.ByteString]
+    -- Options
+  , discPrintCrucible :: Bool
+  }
+
+discoverConfig :: Opt.Parser DiscoverConfig
+discoverConfig =
+  DiscoverConfig
+  <$> MDCU.binOpt
+  <*> Opt.many (Opt.strArgument (Opt.help "function name" <> Opt.metavar "SYMBOL"))
```

Comment:
It might be worth mentioning in the help text that the function names must be UTF-8–encoded.

Reply:
What makes you say that? They are stored as `ByteString`s.

Reply:
Yes, but `displayCfgs` decodes them to `Text` under the assumption that they're UTF-8–encoded.

Reply:
Yes, but it does so with [`lenientDecode`](https://hackage.haskell.org/package/text-2.1.2/docs/Data-Text-Encoding-Error.html#v:lenientDecode), so I don't think I would say they *must* be UTF-8 encoded.

Reply:
> I don't think I would say they must be UTF-8 encoded.

On the other hand, you'll get very surprising results if you give it a UTF-16–encoded string like `roșu`. (It won't cause your program to crash, but it might make you scratch your head wondering why `roșu` turned into `rou`.)

In any case, this isn't that big of a deal, but I usually like to draw attention to this kinds of potential pitfalls.

Reply:
Totally - I'm trying to grok whether this is indeed a pitfall or if it's actually OK. Very happy to document it if it does end up being confusing. The docs for `lenientDecode` say:

> Replace an invalid input byte with the Unicode replacement character U+FFFD.

[Wikipedia says](https://en.wikipedia.org/wiki/Specials_(Unicode_block)):

> U+FFFD � replacement character used to replace an unknown, unrecognised, or unrepresentable character

To me, these together indicate that your example `roșu`, if it were UTF-16 encoded, would print as `ro�u` or the like, which IMO reasonably clearly indicates an encoding error. Or actually, since UTF-16 isn't even ASCII-compatible, perhaps it would just print as `����`.

Another option would be to print a warning in the case that the function name actually wasn't completely valid UTF-8.

Reply:
Added a note that symbol names are expected but not required to be UTF-8.

Reply:
It likely depends on what locale you're using when printing, but for me, the following happens:

```
> Data.Text.IO.putStrLn $ decodeUtf8With lenientDecode $ encodeUtf16LE "roșu"
rou
```

> Another option would be to print a warning in the case that the function name actually wasn't completely valid UTF-8.

No need to go out of your way for this—I think just leaving a comment somewhere near the place where they're parsed is fine.

Reply:
> It likely depends on what locale you're using when printing, but for me, the following happens:

TIL, thanks for the example! I will try to remember to be more careful about this in the future.